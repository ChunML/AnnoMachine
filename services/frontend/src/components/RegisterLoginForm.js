import React from 'react';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import validator from 'validator';
import { formRules } from './utils/form-rules';

class RegisterLoginForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      username: '',
      password: '',
      valid: false,
    };

    this.handleInputChange = this.handleInputChange.bind(this);
    this.handleButtonClick = this.handleButtonClick.bind(this);
  }

  validateForm() {
    const { username, password } = this.state;

    formRules[0].valid = !!validator.isLength(username, { min: 5 });
    formRules[1].valid = !!validator.isHalfWidth(username);
    formRules[2].valid = !!validator.isLength(password, { min: 8 });

    this.setState({
      valid: formRules.reduce((res, cur) => ({ valid: res.valid && cur.valid }))
        .valid,
    });
  }

  handleInputChange(e) {
    const { name, value } = e.target;
    this.setState(
      {
        [name]: value,
      },
      () => this.validateForm()
    );
  }

  handleButtonClick(e) {
    e.preventDefault();
    const { onButtonClick, formType } = this.props;

    onButtonClick(this.state, formType);
  }

  render() {
    const { isAuthenticated, formType } = this.props;
    const { username, password, valid } = this.state;
    if (isAuthenticated) {
      return <Redirect to="/" />;
    }

    return (
      <div className="container column" style={{ margin: '30px auto' }}>
        <ul style={{ listStyle: 'none', display: 'block' }}>
          {formRules.map(rule => (
            <li
              key={rule.id}
              style={{ color: rule.valid ? '#11CC11' : '#EE1233' }}
            >
              {rule.valid && 'âœ”'}
              {rule.message}
            </li>
          ))}
        </ul>
        <div className="form">
          <div className="input-field">
            <input
              name="username"
              type="text"
              placeholder="Enter a username"
              value={username}
              onChange={this.handleInputChange}
            />
          </div>
          <div className="input-field">
            <input
              name="password"
              type="password"
              placeholder="Enter a secure password"
              value={password}
              onChange={this.handleInputChange}
            />
          </div>
          <button
            type="submit"
            className={`button${valid ? ' primary' : ''}`}
            onClick={this.handleButtonClick}
            disabled={!valid}
          >
            {formType === 'register' ? 'Register' : 'Log In'}
          </button>
        </div>
      </div>
    );
  }
}

RegisterLoginForm.propTypes = {
  isAuthenticated: PropTypes.bool,
  formType: PropTypes.string.isRequired,
  onButtonClick: PropTypes.func.isRequired,
};

RegisterLoginForm.defaultProps = {
  isAuthenticated: false,
};

export default RegisterLoginForm;
