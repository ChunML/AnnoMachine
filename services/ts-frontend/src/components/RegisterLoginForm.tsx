import React, { useEffect, useState } from "react";
import { Redirect } from "react-router-dom";
import PropTypes from 'prop-types';
import validator from "validator";
import { formRules } from "./utils/form-rules";

interface RegisterLoginFormProps {
  isAuthenticated: boolean;
  formType: string;
  onButtonClick(formData: RegisterLoginFormStates, formType: string): void;
}

interface RegisterLoginFormStates {
  username: string;
  password: string;
  valid: boolean;
}

function RegisterLoginForm(props: RegisterLoginFormProps) {
  const [states, setStates] = useState<RegisterLoginFormStates>({
    username: "",
    password: "",
    valid: false,
  });

  useEffect(() => {
    validateForm();
  }, [states.username, states.password]);

  const validateForm = () => {
    const { username, password } = states;

    formRules[0].valid = !!validator.isLength(username, { min: 5 });
    formRules[1].valid = !!validator.isHalfWidth(username);
    formRules[2].valid = !!validator.isLength(password, { min: 8 });

    const valids = formRules.map((rule) => rule.valid);

    setStates({
      ...states,
      valid: valids.reduce((res, cur) => res && cur),
    });
  };

  const handleInputChange = (e: React.FormEvent<HTMLInputElement>) => {
    const { name, value } = e.currentTarget;
    setStates({
      ...states,
      [name]: value,
    });
  };

  const handleButtonClick = (e: React.MouseEvent<HTMLButtonElement>) => {
    e.preventDefault();
    const { onButtonClick, formType } = props;

    onButtonClick(states, formType);
  };

  const { isAuthenticated, formType } = props;
  const { username, password, valid } = states;
  if (isAuthenticated) {
    return <Redirect to="/" />;
  }

  return (
    <div className="container column" style={{ margin: "30px auto" }}>
      <ul style={{ listStyle: "none", display: "block" }}>
        {formRules.map((rule) => (
          <li
            key={rule.id}
            style={{ color: rule.valid ? "#11CC11" : "#EE1233" }}
          >
            {rule.valid && "âœ”"}
            {rule.message}
          </li>
        ))}
      </ul>
      <div className="form">
        <div className="input-field">
          <input
            name="username"
            type="text"
            placeholder="Enter a username"
            value={username}
            onChange={handleInputChange}
          />
        </div>
        <div className="input-field">
          <input
            name="password"
            type="password"
            placeholder="Enter a secure password"
            value={password}
            onChange={handleInputChange}
          />
        </div>
        <button
          type="submit"
          className={`button${valid ? " primary" : ""}`}
          onClick={handleButtonClick}
          disabled={!valid}
        >
          {formType === "register" ? "Register" : "Log In"}
        </button>
      </div>
    </div>
  );
}

RegisterLoginForm.propTypes = {
  isAuthenticated: PropTypes.bool,
  formType: PropTypes.string.isRequired,
  onButtonClick: PropTypes.func.isRequired,
};

RegisterLoginForm.defaultProps = {
  isAuthenticated: false,
};

export default RegisterLoginForm;
