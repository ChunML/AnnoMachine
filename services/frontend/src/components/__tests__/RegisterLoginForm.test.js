import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import RegisterLoginForm from '../RegisterLoginForm';
import '../../setupTests';

describe('Register form', () => {
  const component = (
    <RegisterLoginForm formType="register" onButtonClick={jest.fn()} />
  );

  it('RegisterForm renders properly', () => {
    const wrapper = shallow(component);
    const form = wrapper.find('.form');
    expect(form.length).toBe(1);
    const inputs = form.find('input');
    expect(inputs.length).toBe(2);
    const instance = wrapper.instance();
    instance.handleInputChange({
      target: {
        name: 'username',
        value: 'testName',
      },
    });
    expect(wrapper.state()).toEqual({
      username: 'testName',
      password: '',
      valid: false,
    });
    inputs.at(1).simulate('change', {
      target: {
        name: 'password',
        value: 'ninechars',
      },
    });
    expect(wrapper.state()).toEqual({
      username: 'testName',
      password: 'ninechars',
      valid: true,
    });
    const button = form.find('button');
    expect(button.text()).toEqual('Register');
    const { onButtonClick } = component.props;
    expect(onButtonClick).toHaveBeenCalledTimes(0);
    button.simulate('click', { preventDefault: jest.fn() });
    expect(onButtonClick).toHaveBeenCalledTimes(1);
  });

  it('RegisterForm renders a snapshot properly', () => {
    const tree = renderer.create(component).toJSON;
    expect(tree).toMatchSnapshot();
  });
});

describe('Log In form', () => {
  const component = (
    <RegisterLoginForm formType="login" onButtonClick={jest.fn()} />
  );

  it('RegisterForm renders properly', () => {
    const wrapper = shallow(component);
    const form = wrapper.find('.form');
    expect(form.length).toBe(1);
    const inputs = form.find('input');
    expect(inputs.length).toBe(2);
    const instance = wrapper.instance();
    instance.handleInputChange({
      target: {
        name: 'username',
        value: 'testName',
      },
    });
    expect(wrapper.state()).toEqual({
      username: 'testName',
      password: '',
      valid: false,
    });
    inputs.at(1).simulate('change', {
      target: {
        name: 'password',
        value: 'ninechars',
      },
    });
    expect(wrapper.state()).toEqual({
      username: 'testName',
      password: 'ninechars',
      valid: true,
    });
    const button = form.find('button');
    expect(button.text()).toEqual('Log In');
    const { onButtonClick } = component.props;
    expect(onButtonClick).toHaveBeenCalledTimes(0);
    button.simulate('click', { preventDefault: jest.fn() });
    expect(onButtonClick).toHaveBeenCalledTimes(1);
  });

  it('RegisterForm renders a snapshot properly', () => {
    const tree = renderer.create(component).toJSON;
    expect(tree).toMatchSnapshot();
  });
});

describe('User is already logged in', () => {
  const component = (
    <RegisterLoginForm
      isAuthenticated
      formType="login"
      onButtonClick={jest.fn()}
    />
  );
  it('Form renders properly', () => {
    const wrapper = shallow(component);
    expect(wrapper.find('Redirect')).toHaveLength(1);
  });
});
