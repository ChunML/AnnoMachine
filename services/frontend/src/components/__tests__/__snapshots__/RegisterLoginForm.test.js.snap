// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register form RegisterForm renders a snapshot properly 1`] = `[Function]`;

exports[`Register form RegisterForm renders a snapshot properly 2`] = `[Function]`;
